src/gerbil/runtime/mop.ss

```diff
@@ -229,174 +242,112 @@ namespace: #f
+;; : TypeDescriptor -> (OrFalse StructTypeDescriptor)
+(def (base-struct/1 klass)
+  (cond
+   ((struct-type? klass) klass)
+   ((class-type? klass) (##type-super klass))
+   ((not klass) #f)
+   (else (error "Not a class or false" klass))))
+
+;; Which is the most specific struct class if any that both klass1 and klass2 are or inherit from?
+;; : TypeDescriptor TypeDescriptor -> (OrFalse StructTypeDescriptor)
+(def (base-struct/2 klass1 klass2)
+  (def s1 (base-struct/1 klass1))
+  (def s2 (base-struct/1 klass2))
+  (cond
+   ((or (not s1) (and s2 (struct-subtype? s1 s2))) s2)
+   ((or (not s2) (and s1 (struct-subtype? s2 s1))) s1)
+   (else (error "Bad mixin: incompatible struct bases" klass1 klass2 s1 s2))))
+
+;; Which is the most specific struct class if any that all argument classes are or inherit from?
+;; : TypeDescriptor ... -> (OrFalse StructTypeDescriptor)
+(def (base-struct . all-supers)
+  (match all-supers
+    ([] #f)
+    ([x] (base-struct/1 x))
+    ([x y] (base-struct/2 x y))
+    ([x y ...] (foldr base-struct/2 x y))))
+
+;; Find the constructor method name for the TypeDescriptor
+;; : (List TypeDescriptor) -> Symbol
+(def (find-super-ctor super)
+  (let lp ((rest super) (ctor #f))
+    (match rest
+      ([hd . rest]
+       (cond
+        ((type-descriptor-ctor hd)
+         => (lambda (xctor)
+              (if (or (not ctor) (eq? ctor xctor))
+                (lp rest xctor)
+                (error "Conflicting implicit constructors" ctor xctor))))
+        (else (lp rest ctor))))
+      (else ctor))))
+
+;; Given a struct super class (or false if none),
+;; a list of super-classes, and
+;; a list of slots,
+;; return the total number of fields and slots
+;; a list of inherited struct fields in offset order,
+;; a table mapping symbol and keyword names to offset, and
+;; a list of non-struct slots.
+;; : (OrFalse StructTypeDescriptor) (List TypeDescriptor) (List Symbol) \
+;; -> Fixnum (List Symbol) (Table (Or Symbol Keyword) -> Fixnum) (List Symbol)
+(def (compute-class-slots std-super mixins slots)
+  (def std-fields (if std-super (type-descriptor-fields std-super) 0))
+  (def field-list (if std-super (assgetq fields: (type-descriptor-plist std-super)) []))
+  (def slot-table (make-hash-table-eq))
+  (def r-slots [])
+  (def (process-slot slot)
+    (cond
+     ((memq slot field-list) (void)) ;; ignore if already a field
+     ((hash-key? slot-table slot) (void)) ;; ignore if already a slot
+     (else
+      (set! r-slots (cons slot r-slots))
+      (hash-put! slot-table slot std-fields)
+      (hash-put! slot-table (symbol->keyword slot) std-fields)
+      (set! std-fields (1+ std-fields)))))
+  (def process-slots (cut for-each process-slot <>))
+  (for-each (lambda (mixin) (process-slots
+                        (assgetq direct-slots: (type-descriptor-plist mixin) [])))
+            (reverse mixins))
+  (process-slots slots)
+  (values std-fields field-list slot-table (reverse r-slots)))
```

Comment:
let's avoid the internal defines here, same as above for `letrec*`, you can just `let*` this.

Reply:
also, use `fx1+`.